[Run Xinghuan] iP

runxinghuan opened the pull request.

Commit:
Level-0

Commit:
Add echo function and customize its greet and exit functions

Echo repeats commands entered by the user

Customize its greet and exit functions to make it more unique

Commit:
Add "addCommand" and "listCommand" functions

"addCommand" stores the command entered
"listCommand" displays the commands stored

Commit:
Add "markTask" function and change to use a class "Task" for tasks

"markTask" function marks and unmarks a task
according to command entered

Change to use a class to represent tasks because tasks now
have a new condition of "isDone"

Commit:
Tweak the code to comply with CS2113 coding standard

Change the variable "noOfTask" to camelCase

Commit:
Add three types of tasks

Use three subclasses of Task class to represent these types of tasks

"ToDo" is same as a normal task
"Deadline" is a task with a deadline
"Event" is a task with a start date and an end date

Commit:
Test using the I/O redirection technique

Use the technique to semi-automate the testing of Ava

Commit:
Refactor the code to improve its quality

Delete some redunant parts in function "markTask"

Reduce the number of parameter of function "listTask" to one
This is possible after adding "numberOfTasks" variable in class "Task"

owx0130 commented on src/main/java/Ava.java:
```diff
@@ -0,0 +1,107 @@
+            System.out.println("Now you have " + 1 + " task in the list~~~");
+        } else {
+            System.out.println("Now you have " + Task.getNumberOfTasks() + " tasks in the list~~~");
+        }
```
Perhaps these LOCs could be moved to a separate method to ensure SLAP? (consider this for the methods `markTask` and `listTask` as well)

owx0130 commented on src/main/java/Ava.java:
```diff
@@ -0,0 +1,107 @@
+
+public class Ava {
+
+    public static void main(String[] args) {
```
I like how SLAP was well practiced in `main`; method names are also well named with appropriate logical flow

owx0130 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,41 @@
+public class Task {
```
I like the method names in the `Task` class; well named with verbs!

owx0130 commented on src/main/java/Ava.java:
```diff
@@ -0,0 +1,107 @@
+            command = command.replace("unmark ", "");
+            taskChanged = Integer.parseInt(command) - 1;
+            tasks[taskChanged].markAsNotDone();
+            System.out.println("OK, I've marked this task as not done yet:");
```
I see that `command` here is recycled, perhaps another variable could be declared here to avoid reusing variables?

owx0130 commented on src/main/java/Ava.java:
```diff
@@ -0,0 +1,107 @@
+        return task.split("/");
+    }
+
+    public static void addTask(Task[] tasks) {
```
Since `addtask` just consists of print statements, should the method be named to better describe it? An alternative name could be `printAfterAddingTask`, since the action of task adding is actually done in `mainProcess`

owx0130 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+
+    public Deadline(String description, String deadline) {
+        super(description);
+        deadline = deadline.replace("by", "by:");
```
I noticed that `deadline` here is also recycled, perhaps consider my recommendation in the prior comment on `Ava.java` from line 64 to 73?

owx0130 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,17 @@
+
+    public Event(String description, String startDate, String endDate) {
+        super(description);
+        startDate = startDate.replace("from", "from:");
+        endDate = endDate.replace("to", "to:");
```
`startDate` and `endDate` are also recycled here, perhaps consider my recommendation in the prior comment on `Ava.java` from line 64 to 73?

owx0130 commented on src/main/java/Ava.java:
```diff
@@ -0,0 +1,107 @@
+                isExit = true;
+                continue;
+            } else if (task.equals("list")) {
+                listTask(tasks);
+                continue;
+            } else if (task.contains("mark")) {
+                markTask(tasks, task);
+                continue;
+            } else if (task.startsWith("todo")) {
+                task = task.replace("todo ", "");
+                tasks[Task.getNumberOfTasks()] = new ToDo(task);
+            } else if (task.startsWith("deadline")) {
+                String[] taskAndDate = trimTask(task, "deadline ");
+                tasks[Task.getNumberOfTasks()] = new Deadline(taskAndDate[0], taskAndDate[1]);
+            } else if (task.startsWith("event")) {
+                String[] taskAndDate = trimTask(task, "event ");
+                tasks[Task.getNumberOfTasks()] = new Event(taskAndDate[0], taskAndDate[1], taskAndDate[2]);
```
I noticed that there are different levels of abstraction between these else-if blocks. Perhaps you could consider moving the body under the `task.startsWith("todo")`, `task.startsWith("deadline")` and `task.startsWith("event")` blocks to separate methods to enhance readability?

runxinghuan replied:
Thank you for your suggestions! They are very helpful in improving my code quality.

owx0130 reviewed (commented):
Looks good to me overall, your code was easy to read and understand. I noted a few minor coding quality violations which I have recommended solutions for in the comments.

shawnpong commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+        this.deadline = deadline;
+    }
+
+    @Override
```
Great use of the Override annotation to signal method overriding

shawnpong commented on src/main/java/Ava.java:
```diff
@@ -0,0 +1,107 @@
+import java.util.Scanner;
```
I like the good practice of not importing java.util.*, avoids potential namespace collisions!

shawnpong commented on src/main/java/Ava.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public static void greet() {
+        printLine();
```
Maybe you could use "\n" to create a newline instead of habing to call printLine()?

shawnpong commented on src/main/java/Ava.java:
```diff
@@ -0,0 +1,107 @@
+                String[] taskAndDate = trimTask(task, "event ");
+                tasks[Task.getNumberOfTasks()] = new Event(taskAndDate[0], taskAndDate[1], taskAndDate[2]);
+            }
+            addTask(tasks);
```
Noticed that addTask(tasks) just calls a bunch a print statements, this might be a little misleading to people reading your code as the tasks have already been added to the list. Maybe you could rename this method to be more descriptive?

runxinghuan replied:
Thank you for your suggestion!

Commit:
Use exceptions to handle errors

When incorrect inputs are entered, Ava now prints out error messages

Commit:
Merge branch 'branch-Level-5'

Commit:
Divide classes into packages

Create two packages "ava" and "task", and divide classes accordingly

Commit:
Merge branch 'branch-A-Packages'

Commit:
Use Java Collections classes for tasks

Change from a Task array to a collection class

Commit:
Add support for deleting tasks from the list

The "deleteTask" method added is able to delete tasks

Commit:
Add createFile method in DataFile class

createFile method creates a file when there is not an "ava.txt"

Commit:
Merge commit '2089e8614238d4264cd531c586fb31dd2e5daecc' into branch-Level-7

Commit:
Add DataFile class

The class saves tasks in "data/ava.txt" and
load the file when the chatbot starts up

Commit:
Merge branch 'branch-Level-6'

Commit:
Merge branch 'branch-Level-7'

Commit:
Add "saveTasks" after deleting a task

Ava now saves the tasks after deleting each task

Account for the situation of 0 task when saving the task

Commit:
Package the App as a JAR file

Main class is ava.Ava

Jai2501 commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,232 @@
+            }
+            printAfterAddingTask(tasks);
+            DataFile.saveTasks(tasks);
+        }
```
This method is too long. It is a convention to keep the method short, usually below 30 lines.

This block of statement is huge and nested. Consider breaking and abstracting this. This will significantly improve the readability of your code.

Code Quality Reference: https://nus-cs2113-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting

Jai2501 commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,232 @@
+        printLine();
+        System.out.println("(⊙_⊙)? I'm sorry!!! But I don't know what that means.");
+        printLine();
+    }
```
A better method naming strategy for this would be: `printUnknownExceptionMessage`

Jai2501 commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,232 @@
+        printLine();
+        System.out.println("(⊙_⊙)? You need to specify the date after '/'");
+        printLine();
+    }
```
Similarly here, a better method naming strategy for this would be: `printFormatExceptionMessage`

Jai2501 commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,232 @@
+            break;
+        }
+        printLine();
+    }
```
Similarly here, improve the method name to enhance readability.

Jai2501 commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,232 @@
+        String[] taskAndDate = task.split("/");
+        taskAndDate[0] = taskAndDate[0].trim();
+        switch (type) {
+        case "todo":
```
Try separating logically linked statements by line breaks. This too enhances the readability of code.

```suggestion
        task = task.replace(type, "");
        
        if (task.isEmpty()) {
            throw new EmptyDescriptionException();
        }
        
        String[] taskAndDate = task.split("/");
        taskAndDate[0] = taskAndDate[0].trim();
        
        switch (type) {
        case "todo":
```

Jai2501 commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,232 @@
+    public static int extractNumber(String type, String command) {
+        command = command.replace(type, "");
+        return Integer.parseInt(command) - 1;
+    }
```
`extractNumber` is a bit vague. How about `extractTaskNumber` or `parseTaskNumber` or `getTaskNumber`? This gives more context to the reader.

Jai2501 commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    public static void deleteTask(ArrayList<Task> tasks, String command) {
+        int taskDeleted;
```
I think instead of `taskDeleted`, there is still possibility that it is not, depending on the inputs and all. So how about `taskNumberToBeDeleted`, as this gives more context too.

Jai2501 commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,232 @@
+            System.out.println("Please tell me which one to delete? (＾＿－)");
+            printLine();
+        }
+    }
```
This again is a nested code, with try inside another try. Trying adding conditions such that the **Happy Path is Prominent**, that is the path of execution that is going to take place most of the times.

Reference: https://nus-cs2113-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent

Jai2501 commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,232 @@
+            System.out.println("Please tell me which one to mark? (＾＿－)");
+            printLine();
+            return;
+        }
```
Similarly here, you need to work on abstraction and improve this nesting condition.

Jai2501 commented on src/main/java/ava/DataFile.java:
```diff
@@ -0,0 +1,117 @@
+            String endDate = "to" + taskAndDate[2];
+            tasks.add(new Event(taskAndDate[0], startDate, endDate, isCompleted));
+        }
+    }
```
The size of this method needs to be reduced.

Reference: https://nus-cs2113-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods

Jai2501 commented on src/main/java/ava/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,5 @@
+
+public class EmptyDescriptionException extends Exception{
+
+}
```
Empty Class. Please add at least a constructor with valid instantiations/assignments, so if this error is thrown, the description shown is correct.

Else please add a comment explaining when would this error be thrown and why is it empty and nothing initialised.

Jai2501 commented on src/main/java/ava/task/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+    public String toString() {
+        return "[D]" + super.toString() + "(" + this.deadline + ")";
+    }
+}
```
Well written!

Jai2501 commented on src/main/java/ava/task/Task.java:
```diff
@@ -0,0 +1,52 @@
+    public static int getNumberOfTasks() {
+        return numberOfTasks;
+    }
+}
```
Missing EOF. Add an extra empty line at the end of file (EOF), as this indicates the end of the file.

You may configure IntelliJ to add this automatically for you when hit `CTRL/CMD + S`.

Jai2501 reviewed (commented):
Hi Xinghuan!

Good job on the work done so far!

I have reviewed your code and have left some comments. Please go through them and get back to me for any clarifications.

Thanks!

Commit:
Make the Storage class more OOP

Construct a storage object whenever Ava starts

Commit:
Add the TaskList class

Construct a tasks object of type TaskList whenever Ava starts

Commit:
Add the Ui class

Ui deals with interactions with the user

Commit:
Add the Parser class

The class deals with making sense of the user command

Commit:
Make the loadTask method in Storage class more readable

Extract the changeTaskFormat method from the loadTask method

Commit:
Add findTask method in the TaskList class

findTask method finds the tasks containing the keyword entered
printAfterFindingTask method in the Ui class prints out tasks found

Commit:
Add JavaDoc comments to the code

runxinghuan committed:
Merge pull request #2 from runxinghuan/branch-Level-9

Add findTask method in the TaskList class

Commit:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/ava/Parser.java
#	src/main/java/ava/tasklist/TaskList.java
#	src/main/java/ava/ui/Ui.java

Commit:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/ava/Parser.java
#	src/main/java/ava/tasklist/TaskList.java
#	src/main/java/ava/ui/Ui.java

runxinghuan committed:
Merge pull request #1 from runxinghuan/branch-A-JavaDoc

Add JavaDoc comments to the code

Commit:
Add a User Guide

Commit:
Change the format of User Guide

Commit:
Add a line between different features in the User Guide
